Raises error when dims match the variable name

kimbochen opened the pull request:
Fixes #5309 , where a data variable misses when its name collides with its dimensions.

- Added an assertion that checks if all dimension labels are different from the variable name.
- Added a test that checks if an error is raised when a variable name collides with its dimension labels.

Commit:
Added assertion that a variable doesn't have the same name as its dimensions.

Commit:
Fixed code format.

OriolAbril commented on pymc/tests/test_idata_conversion.py:
```diff
@@ -602,6 +602,15 @@ def test_issue_5043_autoconvert_coord_values(self):
             )
             assert isinstance(converter.coords["city"], pd.MultiIndex)
 
+    @pytest.mark.xfail(
```
The test should use a `pytest.raises` context and not be marked as xfail. See https://docs.pytest.org/en/6.2.x/assert.html#assertions-about-expected-exceptions

michaelosthege commented on pymc/model.py:
```diff
@@ -1441,6 +1441,9 @@ def add_random_variable(self, var, dims: Optional[Tuple[Union[str, None], ...]]
             assert all(dim in self.coords or dim is None for dim in dims)
+            assert all(
+                var.name != dim for dim in dims
+            ), f"Variable `{var.name}` has the same name as its dimension label."
```
Since asserts can be turned off globally, and because of coverage calculation, the lines above should be an if statement raising an exception such as ValueError.
I'm sorry to nitpick about this, since clearly you were just following the style of the line above.. Could you change both into an if/raise?

michaelosthege commented on pymc/tests/test_idata_conversion.py:
```diff
@@ -602,6 +602,15 @@ def test_issue_5043_autoconvert_coord_values(self):
+        with pm.Model() as pmodel:
+            pm.ConstantData("time", np.ones(9), dims="time")
+            pm.Normal("a")
+            idata = pm.sample(chains=1)
```
The above two lines are unreachable. You could also call model.register_rv directly

kimbochen replied:
Could you be more specific about which 2 lines are unreachable?
I also looked up the documentation of model.register_rv, but I couldn't figure out how to use it.
Would you mind giving me some additional pointers?

Commit:
Changed xfail to pytest.raise, which asserts a ValueError.

Commit:
Replaced 2 asserts with if/raise statements.

Commit:
Fixed error-raising logic.

kimbochen commented:
Thank you @OriolAbril and @michaelosthege for the detailed review!

- I replaced xfail with pytest.raise, which asserts that a ValueError to be raised.
- In `model.py`, I replaced the two assert statements with if / raise statements.

As for the tests, I had trouble using model.register_rv, so I haven't changed the test yet.

michaelosthege commented:
@kimbochen The change in `model.py` looks good!
Our coverage tool things that it's not yet covered, but it actually is. The coverage just didn't receive the upload because some of the tests failed.

![image](https://user-images.githubusercontent.com/5894642/155850550-6e757dfd-30e1-4eb4-b0fe-a1f3f8bd22d2.png)

So your change actually found bugs in these four tests! 🥳
```
ERROR pymc/tests/test_model_graph.py::TestModelWithDims::test_inputs - ValueE...
ERROR pymc/tests/test_model_graph.py::TestModelWithDims::test_compute_graph
ERROR pymc/tests/test_model_graph.py::TestModelWithDims::test_plates - ValueE...
ERROR pymc/tests/test_model_graph.py::TestModelWithDims::test_graphviz - Valu...
```
Can you fid them by renaming the variables in the test case?

> Could you be more specific about which 2 lines are unreachable?
> I also looked up the documentation of model.register_rv, but I couldn't figure out how to use it.
> Would you mind giving me some additional pointers?
```python
with pm.Model() as pmodel:
    pm.ConstantData("time", np.ones(9), dims="time") # 👈 should raise an error
    pm.Normal("a") # 👈 making this line unreachable because of the error above
```
I think you can use `register_rv` like this:
```python
with pm.Model() as pmodel:
    var = at.as_tensor([1, 2, 3])
    pmodel.register_rv(var, name="mytestvar", dims=("time",))
```

Oh and please update the test to assert the message: `with pytest.raises(ValueError, match="same name as its dimension")`.
To make sure it's not passing because of _any_ ValueError.

Commit:
Removed unreachable statements in var-dim name collision test.

Commit:
Fixed error in tests triggered by var-dim name collision.

kimbochen commented:
@michaelosthege Thank you for the guidance!
I'm glad that my contribution managed to find some bugs.

- I updated the test according to your suggestion: https://github.com/pymc-devs/pymc/blob/9316eb88033a8f989f6db73d69bfe1d366ed03ce/pymc/tests/test_idata_conversion.py#L607-L610
- As for the bugs in the tests, the 4 errors seem to be triggered by the name collision of the `time` variable: https://github.com/pymc-devs/pymc/blob/a3bab7ddc0f4612ad039a7d2755f22fa6c8c0a72/pymc/tests/test_model_graph.py#L107
  - The name of `time` variable is changed to `time`.
  - The following mentions of the variable name are all changed accordingly.

michaelosthege reviewed (approved):
Yes, that's looking good. Thanks @kimbochen !
